File: ZMusic-Plugin/src/main/java/cn/iqianye/mc/zmusic/proto/Toast.java
package cn.iqianye.mc.zmusic.proto;

import cn.iqianye.mc.zmusic.ZMusic;
import cn.iqianye.mc.zmusic.ZMusicBukkit;
import cn.iqianye.mc.zmusic.proto.packet.AdvancementPacket;
import cn.iqianye.mc.zmusic.proto.packet.impl.*;
import org.bukkit.entity.Player;

/**
 * 吐司工具
 *
 * @author 真心
 * @since 2022/6/8 14:15
 */
public class Toast {

    public static void sendToast(Object player, String title) {
        String packageName = ZMusicBukkit.plugin.getServer().getClass().getPackage().getName();
        String nms = packageName.substring(packageName.lastIndexOf('.') + 1);
        Player bukkitPlayer = (Player) player;
        AdvancementPacket packet;
        switch (nms) {
            case "v1_19_R2":
                packet = new AdvancementPacket_1_19_R2(bukkitPlayer, title);
                break;
            case "v1_19_R1":
                packet = new AdvancementPacket_1_19_R1(bukkitPlayer, title);
                break;
            case "v1_18_R2":
                packet = new AdvancementPacket_1_18_R2(bukkitPlayer, title);
                break;
            case "v1_18_R1":
                packet = new AdvancementPacket_1_18_R1(bukkitPlayer, title);
                break;
            case "v1_17_R1":
                packet = new AdvancementPacket_1_17_R1(bukkitPlayer, title);
                break;
            case "v1_16_R3":
                packet = new AdvancementPacket_1_16_R3(bukkitPlayer, title);
                break;
            case "v1_16_R2":
                packet = new AdvancementPacket_1_16_R2(bukkitPlayer, title);
                break;
            case "v1_16_R1":
                packet = new AdvancementPacket_1_16_R1(bukkitPlayer, title);
                break;
            case "v1_15_R1":
                packet = new AdvancementPacket_1_15_R1(bukkitPlayer, title);
                break;
            case "v1_14_R1":
                packet = new AdvancementPacket_1_14_R1(bukkitPlayer, title);
                break;
            case "v1_13_R2":
                packet = new AdvancementPacket_1_13_R2(bukkitPlayer, title);
                break;
            case "v1_12_R1":
                packet = new AdvancementPacket_1_12_R1(bukkitPlayer, title);
                break;
            default:
                ZMusic.log.sendErrorMessage("不支持的NMS版本: " + nms);
                return;
        }

        packet.grant();
        packet.revoke();
    }
}

File: src/main/kotlin/me/zhenxin/zmusic/proto/Toast.kt
package me.zhenxin.zmusic.proto

import me.zhenxin.zmusic.logger
import me.zhenxin.zmusic.platform.bukkitPlugin
import me.zhenxin.zmusic.proto.packet.impl.*
import me.zhenxin.zmusic.utils.colored
import taboolib.common.platform.Platform
import taboolib.common.platform.ProxyPlayer
import taboolib.common.platform.function.runningPlatform

/**
 * 吐司
 *
 * @author 真心
 * @since 2022/6/9 3:45
 */

/**
 * 发送Toast
 * @param msg 信息
 */
fun ProxyPlayer.sendToast(msg: String) {
    if (runningPlatform == Platform.BUKKIT) {
        val pkgName = bukkitPlugin.server.javaClass.`package`.name
        val nms = pkgName.substring(pkgName.lastIndexOf('.') + 1)
        val packet = when (nms) {
            "v1_19_R1" -> AdvancementPacket_1_19_R1(cast(), msg)
            "v1_18_R2" -> AdvancementPacket_1_18_R2(cast(), msg)
            "v1_18_R1" -> AdvancementPacket_1_18_R1(cast(), msg)
            "v1_17_R1" -> AdvancementPacket_1_17_R1(cast(), msg)
            "v1_16_R3" -> AdvancementPacket_1_16_R3(cast(), msg)
            "v1_16_R2" -> AdvancementPacket_1_16_R2(cast(), msg)
            "v1_16_R1" -> AdvancementPacket_1_16_R1(cast(), msg)
            "v1_15_R1" -> AdvancementPacket_1_15_R1(cast(), msg)
            "v1_14_R1" -> AdvancementPacket_1_14_R1(cast(), msg)
            "v1_13_R2" -> AdvancementPacket_1_13_R2(cast(), msg)
            "v1_12_R1" -> AdvancementPacket_1_12_R1(cast(), msg)
            else -> {
                logger.info("&cToast not support this NMS version: $nms".colored())
                return;
            }
        }
        packet.grant()
        packet.revoke()
    }
}